src/main/java/org/opensearch/knn/index/query/KNNQueryFactory.java

```diff
@@ -95,7 +102,14 @@ public static Query create(CreateQueryRequest createQueryRequest) {
                         .rescoreContext(rescoreContext)
                         .build();
             }
-            return createQueryRequest.getRescoreContext().isPresent() ? new NativeEngineKnnVectorQuery(knnQuery) : knnQuery;
+
+            if (createQueryRequest.getRescoreContext().isPresent()) {
+                return new NativeEngineKnnVectorQuery(knnQuery, QUERY_UTILS, isInnerHitQuery);
+            } else if (ENGINES_SUPPORTING_MULTI_VECTORS.contains(knnEngine) && isInnerHitQuery) {
+                return new NativeEngineKnnVectorQuery(knnQuery, QUERY_UTILS, isInnerHitQuery);
+            } else {
+                return knnQuery;
+            }
```

Comment:
It was to isolate disk based vector search as a precaution, we can always call NativeQuery. If we do that we should consider if we need KNNQuery and KNNWeight classes

Reply:
@shatejas at-least for this PR, I would like us to track and simplify this logic and then make be take another PR for removing the `KNNQuery`. Atleast for now I think if we want to remove the `KNNQuery` it will be a big refactor which is completely out of scope for this PR. Open for suggestions here.

Reply:
I can create a separate PR for the change if needed, making it easier to revert in case of any unforeseen issues.

Reply:
when you say a separate PR you mean for the simplification of this condition

Reply:
Yes